Document new linker options and trimming libraries (#23766)
dotnet/docs: docs/core/deploying/preparing-libraries-for-trimming.md

    <TrimmerDefaultAction>link</TrimmerDefaultAction>
    <!-- Optional, for non-default workloads -->
    <SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="path/to/MyLibrary.csproj" />
    <TrimmerRootAssembly Include="MyLibrary" />
  </ItemGroup>

</Project>
```

```dotnetcli
dotnet publish -c Release
```

- Publishing a self-contained app ensures that the library is analyzed in a context where its dependencies are available, so that you are alerted if your library uses any code from dependencies that is incompatible with trimming.

- `TrimmerRootAssembly` ensures that every part of the library is analyzed. This is necessary in case the library has `[AssemblyMetadata("IsTrimmable", "True")]`, which would otherwise let trimming remove the unused library without analyzing it.

- `<TrimmerDefaultAction>link</TrimmerDefaultAction>` ensures that only used parts of dependencies are analyzed. Without this option, you would see warnings originating from _any_ part of a dependency that doesn't set `[AssemblyMetadata("IsTrimmable", "True")]`, including parts that are unused by your library.

> - `<SuppressTrimAnalysisWarnings>link</SuppressTrimAnalysisWarnings>` is implied by `PublishTrimmed` in the .NET SDK. It is only needed if the app targets form factors like Android where the trim analysis warnings are disabled by default. Consult the documentation for your SDK.

Reviewer: Nit: this should be <SuppressTrimAnalysisWarnings>false<SuppressTrimAnalysisWarnings>

Author: Good catch, thanks!

## Resolving trim warnings

The above steps will produce warnings about code that may cause problems when used in a trimmed app. Here are a few examples of the most common kinds of warnings you may encounter, with recommendations for fixing them.

### RequiresUnreferencedCode

```csharp
using System.Diagnostics.CodeAnalysis;

public class MyLibrary
{
    public static void Foo()
    {
        // warning IL2026 : MyLibrary.Foo: Using method 'MyLibrary.Bar' which has
        // 'RequiresUnreferencedCodeAttribute' can break functionality
        // when trimming application code.
        Bar();
    }

    [RequiresUnreferencedCode("Bar does something incompatible with trimming.")]
    static void Bar()
    {
    }
}
```

This means the library calls a method which has explicitly been annotated as incompatible with trimming, using [`RequiresUnreferencedCodeAttribute`](

> https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.requiresunreferencedcodeattribute?view=net-5.0). To get rid of the warning, consider whether `Foo` needs to call `Bar` to do its job. If so, annotate the caller `Foo` with `RequiresUnreferencedCode` as well; this will "bubble up" the warning so that callers of `Foo` get a warning instead. Once you have "bubbled up" the attribute all the way to public APIs (so that these warnings are produced only for public methods, if at all), you are done. Apps which call your library will now get warnings if they call those public APIs, but these will no longer produce warnings like `IL2104: Assembly 'MyLibrary' produced trim warnings`.

Reviewer: Is this the right place to include the DynamicDependency, Descriptor options to solve the RequiresUnreferencedCode warning and be able to use UnconditionalSuppressMessage?

Author: Yes, I think either in this doc or one next to it would be a good place for these. I might do it in a separate PR if I don't get to it soon.

Reviewer 2: When we add this, we should have a "warning" or "advanced" banner on it. It is definitely a "this is something you can do to make things work, but it is really only for people who know precisely what they are doing".

Author: Added a new section with UnconditionalSuppressMessage. I'm having some trouble coming up with examples for DynamicDependency that aren't super contrived. Do you have any suggestions?

Reviewer 2: > Do you have any suggestions?

One example is when serialization doesn't work. 
* https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Runtime.Serialization.Xml/tests/TrimmingTests/EndToEndTest.cs#L13-L15
* https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L177-L182

Author: Thanks! I will add an example for DynamicDependency in a follow-up.

### DynamicallyAccessedMembers

```csharp
using System.Diagnostics.CodeAnalysis;

public class MyLibrary
{
    public static void Foo(Type type)
    {
        // warning IL2067: MyLibrary.Foo(Type): 'type' argument does not satisfy
        // 'DynamicallyAccessedMemberTypes.PublicMethods' in call to 'MyLibrary.UseMethods(Type)'.
        // The parameter 't' of method 'MyLibrary.Foo(Type)' does not have matching annotations.
        UseMethods(type);
    }

    static void UseMethods(
        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
        Type type)
    {
        foreach (var method in type.GetMethods())
        {
            // ...
        }
    }
}
```

> Here, `Foo` is calling a method which takes a `Type` argument that is annotated with a [`DynamicallyAccessedMembers`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?view=net-5.0) requirement. The requirement states that the value passed in as this argument must represent a type whose public methods are available. In this case, you can fix this by adding the same requirement to the parameter of `Foo`. Like above, once you have bubbled up such warnings to public APIs, you are done.

Reviewer: Maybe we should put another piece of code with the code changes for this and the other examples a before/after to be more explicit.
Also, I'm not comfortable with using "Like above, once you have bubbled ..." If we update the document we might point to something else, better to put something like "Like with RequiresUnreferencedCode, once you have bubbled ..."

---
title: Preparing libraries for trimming
description: Learn how to prepare libraries for trimming.
author: sbomer
ms.author: svbomer
ms.date: 08/25/2020
---

# Preparing libraries for trimming

## Trim warnings in apps

When publishing an app, `PublishTrimmed` will produce trim analysis [warnings](trimming-options.md#analysis-warnings) (in .NET 6+) for patterns that are not statically understood to be compatible with trimming, including patterns in your code and in dependencies.

You will encounter detailed warnings originating from your own code and `ProjectReference` dependencies. You may also see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings` for `PackageReference` libraries. This means that the library contained patterns which are not guaranteed to work in the context of the trimmed app, and may result in a broken app. Consider contacting the author to see if the library can be annotated for trimming.

> To resolve warnings originating from the app code, jump ahead to the instructions on [resolving trim warnings](resolving-trim-warnings.md). If you are interested in making your own `ProjectReference` libraries trim friendly, follow the instructions to [enable library trim warnings](enabling-library-trim-warnings).

Reviewer 2: Both links in this line 404 for me.

If your app only uses parts of a library that are compatible with trimming, consider [enabling trimming](trimming-options.md#trim-additional-assemblies) of this library if it is not already being trimmed. This will only produce warnings if your app uses problematic parts of the library. (You can also [show detailed warnings](trimming-options.md#showing-detailed-warnings) for the library to see which parts of it are problematic.)

## Enabling library trim warnings

These instructions show how to enable and resolve static analysis warnings to prepare a library for trimming. Follow these steps if you are authoring a library and either want to proactively make your library trimmable, or have been contacted by app authors who encountered trim warnings from your library.

> Use the .NET 6 SDK for the best experience. It is possible to [show analysis warnings](trimming-options.md#analysis-warnings) in .NET 5, but this will show detailed warnings for every library, including framework libraries. These instructions assume you are using the .NET 6 SDK.

Reviewer 2: ```suggestion
Use the .NET 6 SDK for the best experience. It is possible to [show analysis warnings](trimming-options.md#analysis-warnings) in .NET 5, but not all framework assemblies were annotated correctly in .NET 5. Also, this will show detailed warnings for every library, including framework libraries. These instructions assume you are using the .NET 6 SDK.
```

> Use the .NET 6 SDK for the best experience. It is possible to [show analysis warnings](trimming-options.md#analysis-warnings) in .NET 5, but not all framework assemblies were annotated correctly in .NET 5. Also, this will show detailed warnings for every library, including framework libraries. These instructions assume you are using the .NET 6 SDK.

Reviewer 3: I would just state that users should use .NET 6 - and not go into the differences in behavior (there are way too many of them).

### Incomplete Roslyn analyzer

> During development, you may set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` (in .NET 6+) in your library project to get a _limited_ set of warnings from the Roslyn analyzer. This analyzer is incomplete and should only be used as a convenience. It is important to follow the next steps to ensure that your library is compatible with trimming.

Reviewer 3: This makes us look pretty bad - to soften this a bit I would add that we're working on improving the analyzer.

Reviewer 3: In fact I would probably change this section. (Rewrite in real english please :wink:)

### Enable analysis across toolchain

You should set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` in your library project. This will not have any effect on the output, but it will enable trim analysis across the entire toolchain.

It will enable the Roslyn analyzer for trimming, which is currently not completely, but it's useful for the cases where it works. It provides immediate response from within IDE or build. It doesn't not report all trim analysis warnings and doesn't cover all patterns though. The analyzer should improve over time to provide wider coverage of the trim incompatible patterns.

Author: Reworded, hopefully this is close to what you had in mind.

- `<SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>` is implied by `PublishTrimmed` in the .NET SDK. It is only needed if the app targets form factors like Android where the trim analysis warnings are disabled by default. Consult the documentation for your SDK.

You can also follow the same pattern for multiple libraries, adding them all to the same project as `ProjectReference` and `TrimmerRootAssembly` item to see trim analysis warnings for more than one library at a time, but note that this will warn about dependencies if _any_ of the root libraries use a trim-unfriendly API in a dependency. To see warnings that have to do with only a particular library, reference that library only.

Note that the analysis results depend on the implementation details of your dependencies. If you update to a new version of a dependency, this may introduce analysis warnings if the new version added non-understood reflection patterns, even if there were no API changes. In other words, introducing trim analysis warnings to a library is a breaking change when the library is used with `PublishTrimmed`.

> ## Resolving trim warnings

Reviewer 2: How do you feel about combining @agocke's guidance blog here? https://github.com/mono/linker/pull/1946

There are good things about both descriptions. I'd rather have 1 really good one that combines the best of both, than have 2 different ones with different information.

Author: I think it makes sense to have both a fixed announcement blog post and docs that we can expand over time. At the moment there's a lot of overlap, but I'd like to keep adding to these instructions so that they read more like a reference for authors who really have to dive into the annotations. @agocke thoughts?

I already see one discrepancy, which is that these docs don't yet mention UnconditionalSuppressMessage (I will add it). I'm also happy to take any other suggestions so that the docs contain a superset of the info in the blog post.

Reviewer 2: I'm fine with 2 posts, as long as the official docs (i.e. this document) is better than the blog. The case I don't want to have happen is the blog to be really good, but the official docs being worse.

Reviewer 4: I think this is good for long-term documentation. This looks more detailed, which I think is better for docs, and mine is more narrative, which may be useful for people just starting. I don't see any reason not to keep both up-to-date since the scenarios will be relevant long-term.

```csharp
    // Warn for calls to Foo, but not for Foo's call to Bar.
    [RequiresUnreferencedCode("Calls Bar.")]
    public static void Foo()
    {
        // ...
    }
```

Once you have "bubbled up" the attribute all the way to public APIs (so that these warnings are produced only for public methods, if at all), you are done. Apps which call your library will now get warnings if they call those public APIs, but these will no longer produce warnings like `IL2104: Assembly 'MyLibrary' produced trim warnings`.

### DynamicallyAccessedMembers

```csharp
using System.Diagnostics.CodeAnalysis;

public class MyLibrary
{
    public static void Foo(Type type)
    {
        // warning IL2067: MyLibrary.Foo(Type): 'type' argument does not satisfy
        // 'DynamicallyAccessedMemberTypes.PublicMethods' in call to 'MyLibrary.UseMethods(Type)'.
        // The parameter 't' of method 'MyLibrary.Foo(Type)' does not have matching annotations.
        UseMethods(type);
    }

    static void UseMethods(

>         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]

Reviewer 2: I think it is better to start with a method that straight uses Reflection, teach how to annotate that method with `DynamicallyAccessedMembers`. Then the next step is to show other methods calling that first method, and how to annotate them.

        Type type)
    {
        foreach (var method in type.GetMethods())
        {
            // ...
        }
    }
}
```

> Here, `Foo` is calling a method which takes a `Type` argument that is annotated with a [`DynamicallyAccessedMembers`](https://docs.microsoft.com/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?view=net-5.0) requirement. The requirement states that the value passed in as this argument must represent a type whose public methods are available. In this case, you can fix this by adding the same requirement to the parameter of `Foo`.

Reviewer 3: I would stay away from `Foo` - I know that at some point policheck wasn't happy about it. It also doesn't look "professional", or it looks "lazy". Basically just like we should not use `Foo` or similar simple name in tests - we should make the names descriptive of what they do.

```csharp
    static Type type;

    static void Bar()
    {
        // warning IL2077: MyLibrary.Bar(Type): 'type' argument does not satisfy
        // 'DynamicallyAccessedMemberTypes.PublicMethods' in call to 'MyLibrary.UseMethods(Type)'.
        // The field 'System.Type MyLibrary::type' does not have matching annotations.
        UseMethods(type);
    }
```

Similarly, here the problem is that the field `type` is passed into a parameter with these requinements. You can fix it by adding `DynamicallyAccessedMembers` to the field. This will warn about code that assigns incompatible values to the field instead. Sometimes this process will continue until a public API is annotated, and other times it will end when a concrete type flows into a location with these requirements. For example:

```csharp
    [DynamicallyAccessedMembers(DynamicallyAccessedMembers.PublicMethods)]
    static Type type;

    static void Baz()
    {
        MyLibrary.type = typeof(System.Tuple);
    }
```

In this case the trim analysis will simply keep public methods of `System.Tuple`, and will not produce further warnings.

### Recommendations

> In general, try to avoid reflection if possible, and when using reflection limit it in scope so that it is reachable only from a small part of the library. For example, avoid using non-understood patterns in places like static constructors that will result in the warning propagating to all members of the class.

Reviewer 2: ```suggestion
In general, try to avoid reflection if possible. When using reflection, limit it in scope so that it is reachable only from a small part of the library. For example, avoid using non-understood patterns in places like static constructors that will result in the warning propagating to all members of the class.
```

> In general, try to avoid reflection if possible. When using reflection, limit it in scope so that it is reachable only from a small part of the library. For example, avoid using non-understood patterns in places like static constructors that will result in the warning propagating to all members of the class.

Reviewer 3: I would probably also mention virtual/interface methods as bad places to annotate (as it spreads very wide).

Ensure you are using the .NET 6 SDK for these steps. They will not work correctly in previous versions.

## Enable Roslyn analyzer

During development, set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` (in .NET 6+) in your library project. This will not have any effect on the output, but it will enable trim analysis during build via a Roslyn analyzer.

The Roslyn analyzer is useful for a fast feedback cycle with IDE integration, but is currently incomplete. It doesn't cover all trim analysis warnings, but the set of patterns it understands will improve over time to give more complete coverage. The Roslyn analyzer also isn't able to analyze the implementations of reference assemblies that you depend on. It is important to follow the next steps to ensure that your library is fully compatible with trimming.

### Showing all warnings

To show all analysis warnings for your library, including warnings about dependencies, create a project like the following that references your library, and publish it with `PublishTrimmed`.

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>

>     <OutputType>Exe</OutputType>

Reviewer 5: The description talks about library but output is exe

Author: Clarified by improving the description before the sample

    <TargetFramework>net6.0</TargetFramework>
    <!-- Use a RID of your choice. -->
    <RuntimeIdentifier>linux-x64</RuntimeIdentifier>
    <PublishTrimmed>true</PublishTrimmed>

>     <TrimmerDefaultAction>link</TrimmerDefaultAction>

Reviewer 5: Is this necessary to have it here? Maybe just mention it bellow

Author: I think it's pretty important - added a comment to the sample

>     <!-- Optional, for non-default workloads -->

Reviewer 5: I don't understand what non-default workloads means

Author: Fixed to describe what it does

    <SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="path/to/MyLibrary.csproj" />

>     <TrimmerRootAssembly Include="MyLibrary" />

Reviewer 5: Isn't this the default behavior ?

Author: It is, unless the library is trimmable. Added a comment inline to make it more clear

---
title: Preparing libraries for trimming
description: Learn how to prepare libraries for trimming.
author: sbomer
ms.author: svbomer
ms.date: 08/25/2020
---

# Preparing libraries for trimming

> 

Author: Thanks, added a short intro!

## Trim warnings in apps

When publishing an app, `PublishTrimmed` will produce trim analysis [warnings](trimming-options.md#analysis-warnings) (in .NET 6+) for patterns that are not statically understood to be compatible with trimming, including patterns in your code and in dependencies.

> You will encounter detailed warnings originating from your own code and `ProjectReference` dependencies. You may also see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings` for `PackageReference` libraries. This means that the library contained patterns which are not guaranteed to work in the context of the trimmed app, and may result in a broken app. Consider contacting the author to see if the library can be annotated for trimming.

Reviewer 6: ```suggestion
You will encounter detailed warnings originating from your own code and `ProjectReference` dependencies. You may also see warnings like `warning IL2104: Assembly 'SomeAssembly' produced trim warnings` for `PackageReference` libraries. This warning means that the library contained patterns which are not guaranteed to work in the context of the trimmed app, and may result in a broken app. Consider contacting the author to see if the library can be annotated for trimming.
```

> To resolve warnings originating from the app code, jump ahead to the instructions on [resolving trim warnings](#resolving-trim-warnings). If you are interested in making your own `ProjectReference` libraries trim friendly, follow the instructions to [enable library trim warnings](#enabling-library-trim-warnings).

Reviewer 6: ```suggestion
To resolve warnings originating from the app code, see [resolving trim warnings](#resolve-trim-warnings). If you are interested in making your own `ProjectReference` libraries trim friendly, follow the instructions to [enable library trim warnings](#enable-library-trim-warnings).
```

- `<TrimmerDefaultAction>link</TrimmerDefaultAction>` ensures that only used parts of dependencies are analyzed. Without this option, you would see warnings originating from _any_ part of a dependency that doesn't set `[AssemblyMetadata("IsTrimmable", "True")]`, including parts that are unused by your library.

- `<SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>` is implied by `PublishTrimmed` in the .NET SDK. It is only needed if the app targets form factors like Android where the trim analysis warnings are disabled by default. Consult the documentation for your SDK.

You can also follow the same pattern for multiple libraries, adding them all to the same project as `ProjectReference` and `TrimmerRootAssembly` item to see trim analysis warnings for more than one library at a time, but note that this will warn about dependencies if _any_ of the root libraries use a trim-unfriendly API in a dependency. To see warnings that have to do with only a particular library, reference that library only.

> Note that the analysis results depend on the implementation details of your dependencies. If you update to a new version of a dependency, this may introduce analysis warnings if the new version added non-understood reflection patterns, even if there were no API changes. In other words, introducing trim analysis warnings to a library is a breaking change when the library is used with `PublishTrimmed`.

Reviewer 6: ```suggestion
> [!NOTE]
> The analysis results depend on the implementation details of your dependencies. If you update to a new version of a dependency, this may introduce analysis warnings if the new version added non-understood reflection patterns, even if there were no API changes. In other words, introducing trim analysis warnings to a library is a breaking change when the library is used with `PublishTrimmed`.
```

> ## Resolving trim warnings

Reviewer 6: ```suggestion
## Resolve trim warnings
```

These instructions show how to enable and resolve static analysis warnings to prepare a library for trimming. Follow these steps if you are authoring a library and either want to proactively make your library trimmable, or have been contacted by app authors who encountered trim warnings from your library.

Ensure you are using the .NET 6 SDK for these steps. They will not work correctly in previous versions.

## Enable Roslyn analyzer

> During development, set `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` (in .NET 6+) in your library project. This will not have any effect on the output, but it will enable trim analysis during build via a Roslyn analyzer.

Reviewer 3: Why only during development?
If I want my library to be trim compatible - I should set this property... as simple as that.
